device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
+              this._fsm.transition('authenticated', connectCallback, connectResult);
+            } else {
+              this._amqp.initializeCBS((err) => {
+                if (err) {
+                  /*Codes_SRS_NODE_DEVICE_AMQP_06_008: [If `initializeCBS` is not successful then the client will be disconnected.]*/
+                  this._fsm.transition('disconnecting', connectCallback, getTranslatedError(err, 'AMQP Transport: Could not initialize CBS'));
+                } else {
+                  /*Codes_SRS_NODE_DEVICE_AMQP_06_006: [If `initializeCBS` is successful, `putToken` shall be invoked If `initializeCBS` is successful, `putToken` shall be invoked with the first parameter `audience`, created from the `sr` of the shared access signature, the actual shared access signature, and a callback.]*/
+                  const sasString = this._config.sharedAccessSignature.toString();
+                  this._amqp.putToken(SharedAccessSignature.parse(sasString, ['sr', 'sig', 'se']).sr, sasString, (err) => {
+                    if (err) {
+                      /*Codes_SRS_NODE_DEVICE_AMQP_06_009: [If `putToken` is not successful then the client will be disconnected.]*/
+                      this._fsm.transition('disconnecting', connectCallback, getTranslatedError(err, 'AMQP Transport: Could not authorize with puttoken'));
+                    } else {
+                      this._fsm.transition('authenticated', connectCallback, connectResult);
+                    }
+                  });
+                }
+              });
+            }
+          },
+          connect: () => this._fsm.deferUntilTransition(),
+          disconnect: (disconnectCallback) => this._fsm.transition('disconnecting', disconnectCallback),
+          sendEvent: () => this._fsm.deferUntilTransition(),
+          updateSharedAccessSignature: () => this._fsm.deferUntilTransition(),
+          getTwinReceiver: () => this._fsm.deferUntilTransition(),
+          sendTwinRequest: () => this._fsm.deferUntilTransition(),
+          onDeviceMethod: () => this._fsm.deferUntilTransition(),
+          attachC2DLink: () => this._fsm.deferUntilTransition()
+        },
+        authenticated: {
+          _onEnter: (connectCallback, connectResult) => {
+            connectCallback(null, connectResult);
+          },
+          connect: (connectCallback) => connectCallback(null, new results.Connected()),
+          disconnect: (disconnectCallback) => this._fsm.transition('disconnecting', disconnectCallback),
+          sendEvent: (message, sendCallback) => {
+            let amqpMessage = AmqpMessage.fromMessage(message);
+            amqpMessage.properties.to = this._d2cEndpoint;
+
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_025: [The `sendEvent` method shall create and attach the d2c link if necessary.]*/
+            if (!this._d2cLink) {
+              this._amqp.attachSenderLink(this._d2cEndpoint, null, (err, link) => {
```

Comment:
>attachSenderLink [](start = 25, length = 16)

attachingSenderLink feels like it should be it's own state.  Because the attachSenderLink function is async, the client could do anything while this is happening.  Consider the client which uses Async to call sendEvent many times in parallel.  Each one of those sendEvent actions could end up inside this code at the same time, and I suspect it will cause problems.  (I wholeheartedly disagree with the assertion that you don't have to worry about concurrency because node is single threaded.)

Reply:
that's why we have state machines in the SenderLink and ReceiverLink objects :) the link state machine will be created synchronously, cached and immediately switched to 'attaching'. then all subsequent calls will be directed to that object and handled by it. we should be good here.

---
In reply to: [141131461](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141131461) [](ancestors = 141131461)

Reply:
actually.. you're right, even though it is cached, a new link will be established. I'll investigate.

---
In reply to: [141166984](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141166984) [](ancestors = 141166984,141131461)

Reply:
technically I think it is OK, for 2 reasons:
1/ keeping links unique where necessary is the job of the caller (the various method/twin clients and the AMQP client which maintains the D2C and C2D link and ensure they are unique.
2/ ultimately I want to get rid of the link cache at base layer.

So i'm actually going to by design this until I realize that this is not possible, 

---
In reply to: [141170034](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141170034) [](ancestors = 141170034,141166984,141131461)